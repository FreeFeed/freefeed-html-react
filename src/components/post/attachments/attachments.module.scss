@import '../../../../styles/helvetica/dark-vars.scss';

.attachments a {
  color: #000088;
  text-decoration: none;

  &:hover .original-link__text {
    text-decoration: underline;
  }
}

.container {
  margin-bottom: 1em;
}

.attachment {
  margin-bottom: 0.5em;
}

.attachment--audio {
  margin-bottom: 1em;
}

.audio__player {
  width: 100%;
}

.original-link__container {
  display: flex;
  align-items: flex-start;
  gap: 0.5em;
}

.original-link__icon {
  opacity: 0.75;
  flex: none;
  color: initial;
  margin-top: 0.2em;
}

.original-link__text {
  word-break: break-word;
}

.original-link__size {
  color: initial;
  opacity: 0.5;
}

.original-link__remove {
  flex: none;
  color: #fff;
  background-color: #333;
  border: none;
  display: flex;
  cursor: pointer;
  padding: 0.2em;
  font-size: 1.2em;
  border-radius: 0.35em;
  outline: 3px solid rgba(255, 255, 255, 0.5);
}

.like-a-video {
  margin-bottom: 0.5em;
}

.like-a-video__button {
  border: none;
  display: flex;
  width: 3em;
  height: 3em;
  justify-content: center;
  align-items: center;
  border-radius: 2px;
  background-color: rgba(128, 128, 128, 0.2);
  font-size: 2em;
}

.like-a-video__player {
  max-width: 100%;
  width: 400px;
  height: auto;
  background-color: #eee;
}
