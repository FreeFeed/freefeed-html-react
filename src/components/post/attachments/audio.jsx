import cn from 'classnames';
import { faHeadphones } from '@fortawesome/free-solid-svg-icons';
import { attachmentPreviewUrl } from '../../../services/api';
import { formatFileSize } from '../../../utils';
import style from './attachments.module.scss';
import { OriginalLink } from './original-link';

export function AudioAttachment({ attachment: att, removeAttachment }) {
  const formattedFileSize = formatFileSize(att.fileSize);

  const title =
    [att.meta?.['dc:creator'], att.meta?.['dc:relation.isPartOf'], att.meta?.['dc:title']]
      .filter(Boolean)
      .join(' â€“ ') || att.fileName;

  const titleAndSize = `${title} (${formattedFileSize})`;

  return (
    <div
      role="figure"
      className={cn(style['attachment'], style['attachment--audio'])}
      aria-label={`Audio attachment ${titleAndSize}`}
    >
      <OriginalLink attachment={att} icon={faHeadphones} removeAttachment={removeAttachment}>
        {title}
      </OriginalLink>
      <div>
        <audio
          className={style['audio__player']}
          src={attachmentPreviewUrl(att.id, 'audio')}
          title={titleAndSize}
          preload="none"
          controls
        />
      </div>
    </div>
  );
}
