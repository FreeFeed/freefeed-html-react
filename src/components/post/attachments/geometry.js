export const maxPreviewAspectRatio = 2.5;

export const maxEditingPreviewWidth = 400;
export const maxEditingPreviewHeight = 175;
export const minEditingPreviewWidth = 60;
export const minEditingPreviewHeight = 60;

export function fitIntoBox(att, boxWidth, boxHeight, upscale = false) {
  const [width, height] = [att.previewWidth ?? att.width, att.previewHeight ?? att.height];
  if (!upscale) {
    boxWidth = Math.min(boxWidth, width);
    boxHeight = Math.min(boxHeight, height);
  }
  const wRatio = width / boxWidth;
  const hRatio = height / boxHeight;

  if (wRatio > hRatio) {
    return { width: boxWidth, height: Math.round(height / wRatio) };
  }

  return { width: Math.round(width / hRatio), height: boxHeight };
}

/**
 * @param {number[]} ratios
 * @param {number} containerWidth
 * @param {number} thumbArea
 * @param {number} gap
 * @returns {{width: number, height: number}[][]}
 */
export function getGallerySizes(ratios, containerWidth, thumbArea, gap) {
  let start = 0;
  const lines = [];
  while (start < ratios.length) {
    const line = getGalleryLine(ratios.slice(start), containerWidth, thumbArea, gap);
    lines.push(line);
    start += line.length;
  }
  return lines;
}

/**
 * @param {number[]} ratios
 * @param {number} containerWidth
 * @param {number} thumbArea
 * @param {number} gap
 * @returns {{width: number, height: number}[]}
 */
function getGalleryLine(ratios, containerWidth, thumbArea, gap) {
  let avgRatio = 0;
  let prevHeight = 0;
  let prevDiff = Infinity;
  let n = 0;
  for (const ratio of ratios) {
    n++;
    avgRatio = (avgRatio * (n - 1) + ratio) / n; // Average ratio of one item
    const avgWidth = (containerWidth - gap * (n - 1)) / n; // Average width of one item
    const height = avgWidth / avgRatio;
    const avgArea = height * avgWidth; // Average area of one item
    const diff = Math.log(avgArea / thumbArea);
    if (Math.abs(diff) >= Math.abs(prevDiff)) {
      return ratios
        .slice(0, n - 1)
        .map((r) => ({ width: Math.floor(r * prevHeight), height: prevHeight }));
    }
    prevDiff = diff;
    prevHeight = Math.round(height);
  }

  // Last line
  if (prevDiff > 0.1) {
    const height = Math.round(Math.sqrt(thumbArea / avgRatio));
    return ratios.map((r) => ({ width: Math.floor(r * height), height }));
  }
  return ratios.map((r) => ({ width: Math.floor(r * prevHeight), height: prevHeight }));
}
