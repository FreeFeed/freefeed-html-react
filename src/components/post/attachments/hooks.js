import { useEffect, useMemo, useState } from 'react';
import { useEvent } from 'react-use-event-hook';
import { attachmentPreviewUrl } from '../../../services/api';
import { openLightbox } from '../../../services/lightbox';

const resizeHandlers = new Map();

export function useWidthOf(elRef) {
  const [width, setWidth] = useState(elRef.current?.offsetWidth || 0);
  useEffect(() => {
    const el = elRef.current;
    resizeHandlers.set(el, setWidth);
    const observer = getResizeObserver();
    observer.observe(el);
    return () => {
      resizeHandlers.delete(el);
      observer.unobserve(el);
    };
  }, [elRef]);
  return width;
}

let _observer = null;
function getResizeObserver() {
  if (!_observer) {
    if (globalThis.ResizeObserver) {
      _observer = new globalThis.ResizeObserver((entries) => {
        for (const entry of entries) {
          resizeHandlers.get(entry.target)?.(entry.contentRect.width);
        }
      });
    } else {
      _observer = {
        observe() {},
        unobserve() {},
      };
    }
  }
  return _observer;
}

export function useLightboxItems(attachments, postId) {
  return useMemo(
    () =>
      attachments.map((a) => ({
        ...(a.mediaType === 'image'
          ? { type: 'image', src: attachmentPreviewUrl(a.id, 'image') }
          : {
              type: a.meta?.inProgress ? 'in-progress' : 'video',
              videoSrc: attachmentPreviewUrl(a.id, 'video'),
              msrc: attachmentPreviewUrl(a.id, 'image'),
              meta: a.meta ?? {},
              duration: a.duration ?? 0,
            }),
        originalSrc: attachmentPreviewUrl(a.id, 'original'),
        width: a.previewWidth ?? a.width,
        height: a.previewHeight ?? a.height,
        pid: `${postId?.slice(0, 8) ?? 'new-post'}-${a.id.slice(0, 8)}`,
      })),
    [attachments, postId],
  );
}

export function useItemClickHandler(lightboxItems) {
  return useEvent((e) => {
    if (e.button !== 0 || e.altKey || e.ctrlKey || e.metaKey || e.shiftKey) {
      return;
    }
    e.preventDefault();
    const { currentTarget: el } = e;
    const index = lightboxItems.findIndex((i) => i.pid === el.dataset.pid);
    openLightbox(index, lightboxItems, el.target);
  });
}
