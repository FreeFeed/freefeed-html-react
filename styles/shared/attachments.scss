// General

.attachments {
  clear: both;
  margin-bottom: 2px;

  a {
    color: #000088;
    text-decoration: none;

    &:hover span {
      text-decoration: underline;
    }
  }

  .remove-attachment {
    display: none;
  }

  // Clearfix (http://www.cssmojo.com/latest_new_clearfix_so_far/)
  &:after {
    content: '';
    display: table;
    clear: both;
  }
}

.image-attachments {
  .attachment {
    position: relative;
    display: inline-block;
    vertical-align: middle;
    border: 1px solid silver;
    padding: 1px;
    margin: 0 8px 8px 0;
    box-sizing: content-box;
    text-align: center;

    img {
      max-width: 525px;
      max-height: 175px;
      @media (max-width: 400px) {
        max-width: 250px;
        height: auto;
     }
    }

    &.surplus {
      display: none;
    }

    a {
      display: block;
      min-width: 32px;
      min-height: 32px;
      line-height: 30px;
    }
  }

  &.single-image .attachment {
    height: auto;
    line-height: normal;
  }

  &.expanded .attachment.surplus {
    display: inline-block;
  }

  .toggle-surplus {
    display: none;
    color: #fad889; // logotype yellow (#f9b616) with lower saturation;
    font-size: 2em;
    text-decoration: none;
    position: relative;
    top: 2px;
  }

  &.has-surplus .toggle-surplus {
    display: inline-block;

    // fa-chevron-circle-right
    &:before {
      content: "\f138";
    }
  }

  &.has-surplus.expanded .toggle-surplus {
    // fa-chevron-circle-left
    &:before {
      content: "\f137";
    }
  }
}

.audio-attachments,
.general-attachments {
  .attachment {
    position: relative;
    display: block;
    margin: 0 8px 8px 0;

    i {
      color: #666666;
      padding: 0 1px;
      margin-right: 4px;
    }
  }
}


// Editing entries

.image-attachments {
  .attachment .remove-attachment {
    display: block;
    position: absolute;
    right: 1px;
    top: 1px;
    cursor: pointer;
    color: #000;
    font-size: 20px;
    background-color: silver;
    opacity: 0.7;
    border-bottom-left-radius: 5px;

    &:hover {
      opacity: 1;
    }
  }

  .attachment.removed {
    border: 2px solid #e33;
    padding: 0;
  }

  .attachment.added {
    border: 2px solid #3d3;
    padding: 0;
  }

  .show-more {
    display: none;
    width: 24px;
    color: #8ab;
    vertical-align: middle;
    margin-bottom: 8px;

    .fa {
      cursor: pointer;
      transform: rotate(-180deg);
      transition: transform 0.3s 0.1s;
    }
  }

  &.needs-folding {
    .show-more {
      display: inline-block;
    }
  }

  &.is-folded .show-more .fa {
    transform: rotate(0);
  }
}

.audio-attachments,
.general-attachments {
  .attachment:hover .remove-attachment {
    display: block;
    position: absolute;
    left: -3px;
    bottom: 0;
    cursor: pointer;
    padding: 3px 4px 3px 5px;
    background-color: #fff;
    border-radius: 2px;

    &:hover {
      background-color: #ddd;
    }
  }

  .attachment.removed {
    background-color: #fbb;
  }

  .attachment.added {
    background-color: #bfb;
  }
}
