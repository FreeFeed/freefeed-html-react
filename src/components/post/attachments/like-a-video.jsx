import { useRef, useState } from 'react';
import { useEvent } from 'react-use-event-hook';
import { faPlayCircle } from '@fortawesome/free-solid-svg-icons';
import { attachmentPreviewUrl } from '../../../services/api';
import { Icon } from '../../fontawesome-icons';
import style from './attachments.module.scss';
import { useStopVideo } from './visual/hooks';

export function LikeAVideo({ attachment: att }) {
  const [isOpened, setIsOpened] = useState(false);

  const handleOpen = useEvent(() => setIsOpened(true));

  const videoRef = useRef(null);

  useStopVideo(videoRef, isOpened);

  if (isOpened) {
    return (
      <div className={style['like-a-video']}>
        <video
          ref={videoRef}
          className={style['like-a-video__player']}
          title={att.fileName}
          autoPlay
          controls
          src={attachmentPreviewUrl(att.id, 'original')}
        />
      </div>
    );
  }

  return (
    <div className={style['like-a-video']}>
      <button
        className={style['like-a-video__button']}
        onClick={handleOpen}
        title="Click to play video"
      >
        <Icon icon={faPlayCircle} />
      </button>
    </div>
  );
}
