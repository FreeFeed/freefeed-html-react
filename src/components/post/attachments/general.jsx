import cn from 'classnames';
import { useEvent } from 'react-use-event-hook';
import { faPaperclip } from '@fortawesome/free-solid-svg-icons';
import { formatFileSize, handleLeftClick } from '../../../utils';
import style from './attachments.module.scss';
import { OriginalLink } from './original-link';
import { LikeAVideo } from './like-a-video';

const videoTypes = {
  mov: 'video/quicktime',
  mp4: 'video/mp4; codecs="avc1.42E01E"',
  ogg: 'video/ogg; codecs="theora"',
  webm: 'video/webm; codecs="vp8, vorbis"',
};

const supportedVideoTypes = [];
{
  // find video-types which browser supports
  let video = document.createElement('video');
  for (const [extension, mime] of Object.entries(videoTypes)) {
    if (video.canPlayType(mime) === 'probably') {
      supportedVideoTypes.push(extension);
    }
  }
  video = null;
}

export function GeneralAttachment({ attachment: att, removeAttachment }) {
  const { inProgress = false } = att.meta ?? {};

  const handleClick = useEvent(
    handleLeftClick((e) => {
      if (this.props.meta?.inProgress) {
        e.preventDefault();
        alert('This file is still being processed');
      }
    }),
  );

  const nameAndSize = `${att.fileName} (${inProgress ? 'processing...' : formatFileSize(att.fileSize)})`;

  const extension = att.fileName.split('.').pop().toLowerCase();

  return (
    <div
      role="figure"
      aria-label={`Attachment ${nameAndSize}`}
      className={cn(style['attachment'], style['attachment--general'])}
    >
      {supportedVideoTypes.includes(extension) && <LikeAVideo attachment={att} />}
      <OriginalLink
        attachment={att}
        icon={faPaperclip}
        onClick={handleClick}
        removeAttachment={removeAttachment}
      />
    </div>
  );
}
