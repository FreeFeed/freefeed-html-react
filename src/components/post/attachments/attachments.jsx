import cn from 'classnames';
import { useMemo } from 'react';
import { shallowEqual, useSelector } from 'react-redux';
import { pluralForm } from '../../../utils';
import ErrorBoundary from '../../error-boundary';
import { GeneralAttachment } from './general';
import { AudioAttachment } from './audio';
import style from './attachments.module.scss';
import { VisualContainer } from './visual/container';

export function Attachments({
  attachmentIds,
  isNSFW,
  isExpanded,
  removeAttachment,
  reorderImageAttachments,
  postId,
}) {
  const attachments = useSelector(
    (state) => (attachmentIds || []).map((id) => state.attachments[id]).filter(Boolean),
    shallowEqual,
  );

  const [visualAttachments, audialAttachments, generalAttachments] = useMemo(() => {
    const visual = [];
    const audial = [];
    const general = [];
    for (const a of attachments) {
      if (a.mediaType === 'image' || a.mediaType === 'video') {
        visual.push(a);
      } else if (a.mediaType === 'audio') {
        audial.push(a);
      } else {
        general.push(a);
      }
    }

    return [visual, audial, general];
  }, [attachments]);

  if (attachments.length === 0) {
    return null;
  }

  return (
    <div
      className={style['attachments']}
      role="region"
      aria-label={pluralForm(attachments.length, 'attachment')}
    >
      <ErrorBoundary>
        {visualAttachments.length > 0 && (
          <VisualContainer
            attachments={visualAttachments}
            isNSFW={isNSFW}
            isExpanded={isExpanded}
            removeAttachment={removeAttachment}
            reorderImageAttachments={reorderImageAttachments}
            postId={postId}
          />
        )}
        {audialAttachments.length > 0 && (
          <div className={cn(style['container'], style['container--audio'])}>
            {audialAttachments.map((a) => (
              <AudioAttachment key={a.id} removeAttachment={removeAttachment} attachment={a} />
            ))}
          </div>
        )}
        {generalAttachments.length > 0 && (
          <div className={cn(style['container'], style['container--general'])}>
            {generalAttachments.map((a) => (
              <GeneralAttachment key={a.id} removeAttachment={removeAttachment} attachment={a} />
            ))}
          </div>
        )}
      </ErrorBoundary>
    </div>
  );
}
