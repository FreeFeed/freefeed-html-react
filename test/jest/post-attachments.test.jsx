/* global describe, it, expect */
import { render } from '@testing-library/react';

import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { Attachments } from '../../src/components/post/attachments/attachments';

function renderPostAttachments(attachments = []) {
  const state = { attachments: attachments.reduce((p, a) => ({ ...p, [a.id]: a }), {}) };
  const dummyReducer = (state) => state;
  const store = createStore(dummyReducer, state);
  return render(
    <Provider store={store}>
      <Attachments attachmentIds={attachments.map((a) => a.id)} />
    </Provider>,
  );
}

describe('PostAttachments', () => {
  it('Renders an empty attachments container', () => {
    const { asFragment } = renderPostAttachments();
    expect(asFragment()).toMatchSnapshot();
  });

  it('Displays all post attachment types', () => {
    const image1 = {
      id: 'im1',
      mediaType: 'image',
      fileName: 'CAT.jpg',
      fileSize: 200000,
      previewTypes: ['image'],
      width: 2000,
      height: 1500,
    };

    const image2 = {
      id: 'im2',
      mediaType: 'image',
      fileName: 'food.jpg',
      fileSize: 2000,
      previewTypes: ['image'],
      width: 2000,
      height: 1500,
    };

    const video1 = {
      id: 'vi1',
      mediaType: 'video',
      fileName: 'sunrise.mp4',
      fileSize: 123456789,
      previewTypes: ['image', 'video'],
      duration: 123,
      width: 1920,
      height: 1080,
    };

    const audio1 = {
      id: 'au1',
      mediaType: 'audio',
      fileName: 'wonderwall.mp3',
      fileSize: 1234567,
      previewTypes: ['audio'],
      duration: 300,
      meta: {
        'dc:creator': 'Oasis',
        'dc:title': 'Wonderwall',
      },
    };

    const general1 = {
      id: 'ge1',
      mediaType: 'general',
      fileName: 'rfc.pdf',
      fileSize: 50000,
      previewTypes: [],
    };

    const { asFragment } = renderPostAttachments([video1, image1, general1, image2, audio1]);
    expect(asFragment()).toMatchSnapshot();
  });
});
