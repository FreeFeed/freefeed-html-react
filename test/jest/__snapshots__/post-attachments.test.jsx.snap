// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PostAttachments > Displays all post attachment types 1`] = `
<DocumentFragment>
  <div
    aria-label="5 attachments"
    class="attachments"
    role="region"
  >
    <div
      style="--gap: 8px;"
    >
      <div
        class="container container--visual"
      >
        <div
          class="row row--stretched"
        >
          <a
            class="link"
            data-pid="new-post-vi1"
            href="https://candy.freefeed.net/v4/attachments/vi1/original?redirect="
            role="figure"
            style="width: 337px; height: 190px;"
            target="_blank"
            title="sunrise.mp4 (117.7 MiB, 1920×1080px)"
          >
            <img
              alt="Video attachment sunrise.mp4"
              aria-hidden="true"
              class="image"
              height="190"
              id="new-post-vi1"
              loading="lazy"
              src="https://candy.freefeed.net/v4/attachments/vi1/image?redirect=&width=337&height=190"
              width="337"
            />
            <video
              alt="Video attachment sunrise.mp4"
              class="video"
              disablepictureinpicture=""
              height="190"
              loading="lazy"
              loop=""
              playsinline=""
              poster="https://candy.freefeed.net/v4/attachments/vi1/image?redirect=&width=337&height=190"
              preload="none"
              src="https://candy.freefeed.net/v4/attachments/vi1/video?redirect=&width=337&height=190"
              width="337"
            />
            <div
              class="overlay overlay--info"
            >
              <svg
                aria-hidden="true"
                class="fa-icon fa-icon-fas-play"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
                  fill="currentColor"
                />
              </svg>
              2:03
            </div>
          </a>
          <a
            class="link"
            data-pid="new-post-im1"
            href="https://candy.freefeed.net/v4/attachments/im1/original?redirect="
            role="figure"
            style="width: 253px; height: 190px;"
            target="_blank"
            title="CAT.jpg (195.3 KiB, 2000×1500px)"
          >
            <img
              alt="Image attachment CAT.jpg"
              aria-hidden="false"
              class="image"
              height="190"
              id="new-post-im1"
              loading="lazy"
              src="https://candy.freefeed.net/v4/attachments/im1/image?redirect=&width=253&height=190"
              width="253"
            />
          </a>
          <div
            class="fold__box"
          >
            <button
              class="fold__icon"
              title="Show all (3)"
            >
              <svg
                aria-hidden="true"
                class="fa-icon fa-icon-fas-chevron-circle-right"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 8c137 0 248 111 248 248S393 504 256 504 8 393 8 256 119 8 256 8zm113.9 231L234.4 103.5c-9.4-9.4-24.6-9.4-33.9 0l-17 17c-9.4 9.4-9.4 24.6 0 33.9L285.1 256 183.5 357.6c-9.4 9.4-9.4 24.6 0 33.9l17 17c9.4 9.4 24.6 9.4 33.9 0L369.9 273c9.4-9.4 9.4-24.6 0-34z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="container container--audio"
    >
      <div
        aria-label="Audio attachment Oasis – Wonderwall (1.2 MiB)"
        class="attachment attachment--audio"
        role="figure"
      >
        <div
          class="original-link__container"
        >
          <svg
            aria-hidden="true"
            class="original-link__icon fa-icon fa-icon-fas-headphones"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 32C114.52 32 0 146.496 0 288v48a32 32 0 0 0 17.689 28.622l14.383 7.191C34.083 431.903 83.421 480 144 480h24c13.255 0 24-10.745 24-24V280c0-13.255-10.745-24-24-24h-24c-31.342 0-59.671 12.879-80 33.627V288c0-105.869 86.131-192 192-192s192 86.131 192 192v1.627C427.671 268.879 399.342 256 368 256h-24c-13.255 0-24 10.745-24 24v176c0 13.255 10.745 24 24 24h24c60.579 0 109.917-48.098 111.928-108.187l14.382-7.191A32 32 0 0 0 512 336v-48c0-141.479-114.496-256-256-256z"
              fill="currentColor"
            />
          </svg>
          <div
            class="original-link__text"
          >
            <a
              class="original-link"
              href="https://candy.freefeed.net/v4/attachments/au1/original?redirect="
              target="_blank"
            >
              <span
                class="original-link__text"
              >
                Oasis – Wonderwall
              </span>
               
              <span
                class="original-link__size"
              >
                1.2 MiB
              </span>
            </a>
          </div>
        </div>
        <div>
          <audio
            class="audio__player"
            controls=""
            preload="none"
            src="https://candy.freefeed.net/v4/attachments/au1/audio?redirect="
            title="Oasis – Wonderwall (1.2 MiB)"
          />
        </div>
      </div>
    </div>
    <div
      class="container container--general"
    >
      <div
        aria-label="Attachment rfc.pdf (48.8 KiB)"
        class="attachment attachment--general"
        role="figure"
      >
        <div
          class="original-link__container"
        >
          <svg
            class="original-link__icon fa-icon fa-icon-fas-paperclip"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              xlink:href="#fa-icon-fas-paperclip"
            />
          </svg>
          <div
            class="original-link__text"
          >
            <a
              class="original-link"
              href="https://candy.freefeed.net/v4/attachments/ge1/original?redirect="
              target="_blank"
            >
              <span
                class="original-link__text"
              >
                rfc.pdf
              </span>
               
              <span
                class="original-link__size"
              >
                48.8 KiB
              </span>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PostAttachments > Renders an empty attachments container 1`] = `<DocumentFragment />`;
